h5py/tests/test_dtype.py

```diff
@@ -394,37 +394,77 @@ class TestB8(TestCase):
 
     def test_b8_bool(self):
         arr1 = np.array([False, True], dtype=bool)
-        self._test_b8(arr1)
-        self._test_b8(arr1, dtype=np.uint8)
+        self._test_b8(
+            arr1,
+            expected_default_cast_dtype=np.uint8
+        )
+        self._test_b8(
+            arr1,
+            expected_default_cast_dtype=np.uint8,
+            cast_dtype=np.uint8
+        )
 
     def test_b8_bool_compound(self):
         arr1 = np.array([(False,), (True,)], dtype=np.dtype([('x', '?')]))
-        self._test_b8(arr1)
-        self._test_b8(arr1, dtype=np.dtype([('x', 'u1')]))
+        self._test_b8(
+            arr1,
+            expected_default_cast_dtype=np.dtype([('x', 'u1')])
+        )
+        self._test_b8(
+            arr1,
+            expected_default_cast_dtype=np.dtype([('x', 'u1')]),
+            cast_dtype=np.dtype([('x', 'u1')])
+        )
 
     def test_b8_bool_compound_nested(self):
         arr1 = np.array(
             [(True, (True, False)), (True, (False, True))],
             dtype=np.dtype([('x', '?'), ('y', [('a', '?'), ('b', '?')])]),
         )
-        self._test_b8(arr1)
         self._test_b8(
             arr1,
-            dtype=np.dtype([('x', 'u1'), ('y', [('a', 'u1'), ('b', 'u1')])]),
+            expected_default_cast_dtype=np.dtype(
+                [('x', 'u1'), ('y', [('a', 'u1'), ('b', 'u1')])]
+            )
+        )
+        self._test_b8(
+            arr1,
+            expected_default_cast_dtype=np.dtype(
+                [('x', 'u1'), ('y', [('a', 'u1'), ('b', 'u1')])]
+            ),
+            cast_dtype=np.dtype([('x', 'u1'), ('y', [('a', 'u1'), ('b', 'u1')])]),
+        )
+
+    def test_b8_bool_compound_mixed_types(self):
+        arr1 = np.array(
+            [(True, 0.5), (False, 0.2)], dtype=np.dtype([('x','?'), ('y', '<f8')])
+        )
+        self._test_b8(
+            arr1,
+            expected_default_cast_dtype=np.dtype([('x', 'u1'), ('y', '<f8')])
+        )
+        self._test_b8(
+            arr1,
+            expected_default_cast_dtype=np.dtype([('x', 'u1'), ('y', '<f8')]),
+            cast_dtype=np.dtype([('x', 'u1'), ('y', '<f8')])
         )
 
     def test_b8_bool_array(self):
         arr1 = np.array(
             [((True, True, False),), ((True, False, True),)],
             dtype=np.dtype([('x', ('?', (3,)))]),
         )
-        self._test_b8(arr1)
         self._test_b8(
             arr1,
-            dtype=np.dtype([('x', ('?', (3,)))]),
+            expected_default_cast_dtype=np.dtype([('x', ('u1', (3,)))])
+        )
+        self._test_b8(
+            arr1,
+            expected_default_cast_dtype=np.dtype([('x', ('u1', (3,)))]),
+            cast_dtype=np.dtype([('x', ('?', (3,)))]),
         )
 
-    def _test_b8(self, arr1, dtype=None):
+    def _test_b8(self, arr1, expected_default_cast_dtype, cast_dtype=None):
```

Comment:
While we're at it, could we have one test of a longer bitfield type (i.e. anything beyond 8-bit) as well?

Reply:
This is something I've struggled to work out how to do - since we don't have reflection for bitfield types, my understanding is that we're only able to read them, not write them, since the matching `numpy.uint` equivalents will be written as `H5T_NATIVE_UINT` types. The way we circumvent this for the 8 bit case is to use `tables`, which writes it's bools as `H5T_NATIVE_B8`. I did some digging and couldn't find any information about other types that write as larger bitfields - are you able to provide any guidance on how we might be able to test the larger bitfield types?

Reply:
Aha, good point. You've added conversions both ways, so writing uint data into a bitfield dataset shouldn't be a problem, but creating it in the first place is a bit trickier. I think at present, we'd need to use the [low-level API](https://api.h5py.org/). Something like this (untested):

```python
f = h5py.File('foo.h5', 'w')

space = h5py.h5s.create_simple((10,))
dset_id = h5py.h5d.create(f.id, b'data', h5py.h5t.STD_B32LE, space)
dset = h5py.Dataset(dset_id)
```

(We should probably add the STD & NATIVE bitfield types to `h5t.pyx`, in the same way that the equivalent integer types are there. I've assumed that in the example)